src/common/Fixed.ml

```diff
@@ -0,0 +1,236 @@
+  meta data `'a`. That is, a tuple `('a * 'b t Pattern.t)` has two type 
+  variables where as  `'a t` has one.
+  *)
+
+  val rewrite_bottom_up : f:('a t -> 'a t) -> 'a t -> 'a t
+  (** `rewrite_bottom_up` specializes `fold_pattern` so that the result type 
+  `'r` is equal to the type of our fixed-point data structure i.e. `'r = 'a t`. 
+  This also means that the function `f` can be written with our fixed-point type 
+  `'a t` as its argument. *)
+
+  val unfold_pattern : f:('r -> 'a * 'r Pattern.t) -> 'r -> 'a t
+  (** `unfold` builds a fixed-point data structure from the top down. Starting
+  with a user-supplied seed of type `'r`, `unfold` recursively applies the 
+  function `f` yielding a tuple of meta-data and pattern with elements of 
+  type `'r` to which `f` is further applied. 
+  
+  The `unfold` terminates when a pattern element which does not carry an `'r` is 
+  reached.
+
+  As with `fold_pattern` the function `f` returns a tuple of meta-data and 
+  pattern rather than our record type since, in general, `'r =/= 'a t`. *)
+
+  val rewrite_top_down : f:('a t -> 'a t) -> 'a t -> 'a t
+  (** `rewrite_top_down` specializes `unfold` by requiring that `'r = 'a t`. 
+  As a consequence the function `f` accepts our record type `'a t` as its 
+  argument. *)
+end
+
+(** Functor  which creates the fixed-point of the type defined in the `Pattern`
+module argument 
+*)
+module Make (Pattern : Pattern.S) : S with module Pattern := Pattern = struct
+  type 'a t = {pattern: 'a t Pattern.t; meta: 'a}
+  [@@deriving compare, map, fold, hash, sexp]
+
+  let fix (meta, pattern) = {meta; pattern}
+  let unfix {meta; pattern} = (meta, pattern)
+  let pattern_of {pattern; _} = pattern
+  let meta_of {meta; _} = meta
+
+  let rec pp f ppf {pattern; meta} =
+    Fmt.pf ppf {|%a%a|} f meta (Pattern.pp (pp f)) pattern
+
+  include Foldable.Make (struct type nonrec 'a t = 'a t
+
+                                let fold = fold
+  end)
+
+  let rec fold_pattern ~f t =
+    let meta, pattern = unfix t in
+    let pattern' = Pattern.map (fold_pattern ~f) pattern in
+    f (meta, pattern')
+
+  (** For clarity this is written explicitly but is equivalent to 
+  `fold_pattern ~f:(Fn.compose f fix) t` 
+  *)
+  let rec rewrite_bottom_up ~f t =
+    let x = {t with pattern= Pattern.map (rewrite_bottom_up ~f) t.pattern} in
+    f x
+
+  let rec unfold_pattern ~f x =
+    let meta, pattern = f x in
+    let pattern' = Pattern.map (unfold_pattern ~f) pattern in
+    fix (meta, pattern')
+
+  (** For clarity this is written explicitly but it is equivalent to
+  `unfold ~f:(Fn.compose unfix f) x` 
+  *)
+  let rec rewrite_top_down ~f x =
+    let t = f x in
+    {t with pattern= Pattern.map (rewrite_top_down ~f) t.pattern}
+end
+
+(** Nested fixed-point type where an element of the `Pattern` is itself
+a fixed-point type. We use this to represent statements which contain 
+expressions.
+*)
+module type S2 = sig
+  module First : S
```

Comment:
I see why you called Expr First earlier, but I'm not sure we need this level of abstraction... I don't think we're going to have any more two-level types in this compiler. All of them will involve expressions and statements. Are there any gains to be had if we assume that prediction is true?

Reply:
I'm not sure what you mean here - are you suggesting changing the name or getting rid of the fixed point type for statements?